From b4250c497c8e520723e2aa183794b69877e32155 Mon Sep 17 00:00:00 2001
From: Elizabeth Figura <zfigura@codeweavers.com>
Date: Thu, 18 Dec 2025 13:56:17 -0600
Subject: [PATCH 39/50] wined3d: Only invalidate enabled clip planes.

When the view matrix or vertex shader changes, that is.

And, to compensate, invalidate clip planes when the clip plane mask changes.

Wine-Bug: https://bugs.winehq.org/show_bug.cgi?id=59013
Signed-off-by: sewn <sewn@disroot.org>
---
 dlls/wined3d/stateblock.c | 23 +++++++++++++++++------
 1 file changed, 17 insertions(+), 6 deletions(-)

diff --git a/dlls/wined3d/stateblock.c b/dlls/wined3d/stateblock.c
index af192cf..aa3a710 100644
--- a/dlls/wined3d/stateblock.c
+++ b/dlls/wined3d/stateblock.c
@@ -1715,6 +1715,19 @@ void CDECL wined3d_stateblock_set_render_state(struct wined3d_stateblock *stateb
         case WINED3D_RS_CLIPPING:
         case WINED3D_RS_CLIPPLANEENABLE:
             stateblock->changed.extra_vs_args = 1;
+            /* Clip planes are affected by the view matrix. However in order to
+             * avoid unnecessarily invalidating clip planes every time the view
+             * matrix changes (which is itself frequent, but clipping might not
+             * even be enabled) we only invalidate the currently active clip
+             * planes.
+             *
+             * This means that we need to recalculate clip planes when the clip
+             * plane mask changes. If we changed the view matrix while clip
+             * planes were disabled, we need to recalculate them now that
+             * they've been enabled, since they weren't recalculated at the
+             * time. */
+            if (stateblock->stateblock_state.rs[WINED3D_RS_CLIPPING])
+                stateblock->changed.clipplane |= stateblock->stateblock_state.rs[WINED3D_RS_CLIPPLANEENABLE];
             break;
 
         case WINED3D_RS_ALPHAFUNC:
@@ -3075,7 +3088,8 @@ void CDECL wined3d_device_apply_stateblock(struct wined3d_device *device,
     {
         /* Clip planes are affected by the view matrix, but only if not using
          * vertex shaders. */
-        changed->clipplane = wined3d_mask_from_size(WINED3D_MAX_CLIP_DISTANCES);
+        if (state->rs[WINED3D_RS_CLIPPING])
+            changed->clipplane |= state->rs[WINED3D_RS_CLIPPLANEENABLE];
     }
 
     for (start = 0; ; start = range.offset + range.size)
@@ -3665,8 +3679,8 @@ void CDECL wined3d_device_apply_stateblock(struct wined3d_device *device,
     {
         if (wined3d_bitmap_is_set(changed->transform, WINED3D_TS_VIEW))
         {
-            changed->lights = 1;
-            changed->clipplane = wined3d_mask_from_size(WINED3D_MAX_CLIP_DISTANCES);
+            if (state->rs[WINED3D_RS_CLIPPING])
+                changed->clipplane |= state->rs[WINED3D_RS_CLIPPLANEENABLE];
         }
 
         if (wined3d_bitmap_is_set(changed->transform, WINED3D_TS_PROJECTION) || changed->position_transformed)
@@ -3674,9 +3688,6 @@ void CDECL wined3d_device_apply_stateblock(struct wined3d_device *device,
                     WINED3D_PUSH_CONSTANTS_VS_FFP, WINED3D_SHADER_CONST_FFP_PROJ,
                     offsetof(struct wined3d_ffp_vs_constants, projection_matrix),
                     sizeof(state->transforms[WINED3D_TS_PROJECTION]), &state->transforms[WINED3D_TS_PROJECTION]);
-
-        /* Clip planes are affected by the view matrix. */
-        changed->clipplane = wined3d_mask_from_size(WINED3D_MAX_CLIP_DISTANCES);
     }
 
     if (changed->indices)
-- 
2.52.0

